certbot-nginx/certbot_nginx/configurator.py

```diff
@@ -565,6 +569,39 @@ def _has_certbot_redirect(self, vhost, domain):
         test_redirect_block = _test_block_from_block(_redirect_block_for_domain(domain))
         return vhost.contains_list(test_redirect_block)
 
+    def _set_http_header(self, domain, header_substring):
+        """Enables header identified by header_substring on domain.
+
+        If the vhost is listening plaintextishly, separates out the relevant
+        directives into a new server block, and only add header directive to
+        HTTPS block.
+
+        :param str domain: the domain to enable header for.
+        :param str header_substring: String to uniquely identify a header.
+                        e.g. Strict-Transport-Security, Upgrade-Insecure-Requests
+        :returns: Success
+        :raises .errors.PluginError: If no viable HTTPS host can be created or
+            set with header header_substring.
+        """
+        vhost = self.choose_vhost(domain)
+        if vhost is None:
+            raise errors.PluginError(
+                "Unable to find corresponding HTTPS host for enhancement.")
+
+        if vhost.has_header(header_substring):
```

Comment:
and then probably `for vhost in vhosts` because if `domain` is a wildcard domain we allow multiple vhosts to be chosen